### ISSUE TITLE
adscert 2.0 release pipeline proposal

### ISSUE THREAD

**Issue by shahinrahbariasl** [IABTechLab/adscert#15]
### Objective
Define release pipeline for ads.cert 2.0

_Note:_ This doc does not cover details of each stage of the pipeline.

### Guiding principles and Goals

- Automate release process
- Support adding multiple stages to the pipeline (build, test, vulnerability/security check, etc.)
- Build and publish binaries if applicable
- Build and publish versions and tag releases
- Create CHANGELOG entries on releases
- Modify README and publish go docs automatically on releases
- Publish deployment docs
- Configurable workflows based on different branches (main, staging, develop, etc.)

Nice to have:
- Push Slack deployment notifications
- Deploy Docker images to Docker Hub if applicable

### Proposal
Since ads.cert is going to be hosted on a public github repository, we can leverage github actions and workflows to build a CI/CD pipeline since actions integrate seamlessly with github and source is also open source. Some advantages and downsides are:

### Advantages

- Ease of setup (no installation is required).
- Tight integration with github.
- Asynchronous CI/CD
- Support for many languages and frameworks (including go)
- The individual actions in a workflow are isolated by default. We can use a completely different computing environment for different stages like compilation and testing. For example Travis CI (and most other "traditional" CIs) would run all stages in the same computing environment.

### Disadvantages

- No native cache (need to implement if needed)
- GitHub actions are just consecutive docker runs. Very easy to reason about and debug. But it is a single platform, just whatever you can run inside docker. That seems unlikely to change, but might be an acceptable limitation (we can always add an action to call other cross-platform CI/CD services).

### Alternative
The most reliable and popular options are Travis CI and Jenkis but they are not free. Task are also synchronous and they need installation.

### Setting up Workflows
Workflows are custom automated processes that we can include in the repo to build, test, and deploy the code. We can have one or more workflows running in parallel. Workflows need be stored in `.github/workflows`.

### Overall flow diagram
![adscert-release-pipeline-2](https://user-images.githubusercontent.com/56240400/121726926-34e33c80-cab9-11eb-91e7-f43c23644312.png)

Each workflow has a name. We can then configure events to trigger the workflow

```
on:
 push:
   branches:
   — master
   — staging
   — develop
 pull_request:
   branches:
   — master
   — staging
   — develop
```

Each stage can be run as a job which runs on a runner

```
jobs:
 # The “build” workflow
 build:

runs-on: ubuntu-latest
```

In each job, we can add steps which represent a sequence of tasks to be executed as part of the job

```
steps:
 — uses: actions/checkout@v2
```

Setup environment to use

```
- name: Setup Go
 uses: actions/setup-go@v2
 with:
   go-version: ‘1.14.0’

```
Install dependencies

```
- name: Install dependencies
 run: |
   go version
   go get -u golang.org/x/lint/golint
```

Packaging and building the application

```
- name: Run build
 run: go build .
```

Run vet and lint

```
- name: Run vet & lint
 run: |
   go vet .
   golint .
```

Run tests

```

- name: Run testing
run: cd test && go test -v
```

Slack notifications

```
- name: Send slack notification
 uses: 8398a7/action-slack@v3
 with:
   status: ${{ job.status }}
   fields: repo,message,commit,author,action,eventName,ref,workflow,job,took
 env:
   SLACK_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK_URL }}
 if: always()
```

Deploy Docker image to Docker Hub

```
deploy:
  runs-on: ubuntu-latest
  needs: [build]
  if: ${{ github.ref == ‘refs/heads/master’ && github.event_name == ‘push’ }} 
  steps:
  - uses: actions/checkout@v2
  - name: Deploy to Docker registry
    uses: docker/build-push-action@v1
    with:
      username: ${{ secrets.DOCKER_USERNAME }}
      password: ${{ secrets.DOCKER_PASSWORD }}
      repository: test/gosimple
      tag_with_ref: true
```

### Overview of workflows and running jobs
![Screen Shot 2021-06-11 at 1 32 50 PM](https://user-images.githubusercontent.com/56240400/121727156-87bcf400-cab9-11eb-8084-f757ccd79fd8.png)

### Putting it all together

```
name: CICD

# Controls when the action will run. Triggers the workflow on push or pull request
# events but only for the master branch
on:
  push:
    branches:
    - master
    - staging
    - develop
  pull_request:
    branches:
    - master
    - staging
    - develop

# A workflow run is made up of one or more jobs that can run sequentially or in parallel
jobs:
  # The "build" workflow
  build:
    # The type of runner that the job will run on
    runs-on: ubuntu-latest

    # Steps represent a sequence of tasks that will be executed as part of the job
    steps:
    # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
    - uses: actions/checkout@v2
    
    # Setup Go
    - name: Setup Go
      uses: actions/setup-go@v2
      with:
        go-version: '1.14.0' 
    
    # Install all the dependencies
    - name: Install dependencies
      run: |
        go version
        go get -u golang.org/x/lint/golint
        
    # Run build of the application
    - name: Run build
      run: go build . 
      
    # Run vet & lint on the code
    - name: Run vet & lint
      run: |
        go vet .
        golint .
    
    # Run testing on the code
    - name: Run testing
      run: cd test && go test -v
    
    # Send slack notification
    - name: Send slack notification
      uses: 8398a7/action-slack@v3
      with:
        status: ${{ job.status }}
        fields: repo,message,commit,author,action,eventName,ref,workflow,job,took # selectable (default: repo,message)
      env:
        SLACK_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK_URL }} # required
      if: always() # Pick up events even if the job fails or is canceled.

  # The "deploy" workflow
  deploy:
    # The type of runner that the job will run on
    runs-on: ubuntu-latest
    needs: [build] # Only run this workflow when "build" workflow succeeds
    if: ${{ github.ref == 'refs/heads/master' && github.event_name == 'push' }} # Only run this workflow if it is master branch on push event
    steps:
    - uses: actions/checkout@v2

    # Deploy to Docker registry
    - name: Deploy to Docker registry
      uses: docker/build-push-action@v1
      with:
        username: ${{ secrets.DOCKER_USERNAME }}
        password: ${{ secrets.DOCKER_PASSWORD }}
        repository: test/gosimple
        tag_with_ref: true

```

### Automate changelog and releases notes

```
  changelog:
    name: Update Changelog
    runs-on: ubuntu-latest
    steps:
      - name: Get version from tag
        env:
          GITHUB_REF: ${{ github.ref }}
        run: |
          export CURRENT_VERSION=${GITHUB_TAG/refs\/tags\/v/}
          echo "::set-env name=CURRENT_VERSION::$CURRENT_VERSION"
      - name: Checkout code
        uses: actions/checkout@v2
        with:
          ref: main
      - name: Update Changelog
        uses: heinrichreimer/github-changelog-generator-action@v2.1.1
        with:
          token: ${{ secrets.GITHUB_TOKEN }}
          issues: true
          issuesWoLabels: true
          pullRequests: true
          prWoLabels: true
          addSections: '{"documentation":{"prefix":"**Documentation:**","labels":["documentation"]}}'
      - uses: stefanzweifel/git-auto-commit-action@v4
        with:
          commit_message: Update Changelog for tag ${{ env.CURRENT_VERSION }}
          file_pattern: CHANGELOG.md

  release_notes:
    name: Create Release Notes
    runs-on: ubuntu-latest
    needs: changelog
    steps:
      - name: Get version from tag
        env:
          GITHUB_REF: ${{ github.ref }}
        run: |
          export CURRENT_VERSION=${GITHUB_TAG/refs\/tags\/v/}
          echo "::set-env name=CURRENT_VERSION::$CURRENT_VERSION"
      - name: Checkout code
        uses: actions/checkout@v2
        with:
          ref: main

      - name: Get Changelog Entry
        id: changelog_reader
        uses: mindsers/changelog-reader-action@v1
        with:
          version: ${{ env.CURRENT_VERSION }}
          path: ./CHANGELOG.md

      - name: Create Release
        id: create_release
        uses: actions/create-release@v1
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }} 
        with:
          tag_name: ${{ github.ref }}
          release_name: Release ${{ github.ref }}
          body: ${{ steps.changelog_reader.outputs.log_entry }}
          draft: false
          prerelease: false
```

### Generate go doc 
Go code that we upload to public repositories on github appears automatically on the godoc website. 

We can also define a job for generating docs

```
on: [pull_request]

jobs:
  main:
    runs-on: ubuntu-latest
    name: Example
    steps:
    - name: Checkout
      uses: actions/checkout@v1
    - name: Generate GoDoc
      uses: ktr0731/godoc-action@v0.1.0
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
    - name: Push changes
      uses: ad-m/github-push-action@master
      with:
        github_token: ${{ secrets.GITHUB_TOKEN }}
        branch: gh-pages
```

References

https://docs.github.com/en/actions
https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions
https://github.com/ktr0731/godoc-action
https://pkg.go.dev/github.com/posener/goreadme
https://github.com/renehernandez/camper/blob/main/.github/workflows/release.yml
https://medium.com/swlh/setting-up-github-actions-for-go-project-ea84f4ed3a40

**Comment by shahinrahbariasl** [comment]
@cmlight take a look at this release proposal. The idea is to leverage GitHub Actions to automate the build and deploy process for each PR/MRs.

**Comment by cmlight** [comment]
Thanks, Shahin, for putting this together and chatting with me about it over VC!

@iabmayank @amitshetty Please take a look at this proposal that @shahinrahbariasl wrote up on a proposed release process for the ads.cert OSS.

Mayank: I remember you had mentioned in an email that the Tech Lab has a subscription to Travis CI, and you had also mentioned that using GitHub Actions is another option.  I wanted to see if your team has any preferences from the IABTL perspective for which CI/CD stack to use, in case you want to pursue consistent infrastructure at the organization level?

**Comment by grossjo** [comment]
@iabmayank @amitshetty   Hi,  Any feedback on GitHub actions vs Travis CI?  And in general the proposed pipeline in general.
